Mono.Debugging.Soft/SoftDebuggerSession.cs

```diff
@@ -2640,58 +2640,67 @@ bool LoadPdbFile (string pdbFileName)
-			using (var metadataReader = MetadataReaderProvider.FromPortablePdbStream (fs)) {
-				var pdb = metadataReader.GetMetadataReader ();
-				if (pdb.Documents.Count == 0)//If .pdb has 0 documents consider it invalid
-					return false;
-				var methodMapping = new Dictionary<DocumentHandle, List<SequencePoint []>> (pdb.Documents.Count);
-				foreach (var methodHandle in pdb.MethodDebugInformation) {
-					var method = pdb.GetMethodDebugInformation (methodHandle);
-					if (method.Document.IsNil)
-						continue;
-					List<SequencePoint []> list;
-					if (!methodMapping.TryGetValue (method.Document, out list))
-						methodMapping [method.Document] = list = new List<SequencePoint []> ();
-
-					list.Add (method.GetSequencePoints ().ToArray ());
-				}
-
-				foreach (var documentHandle in pdb.Documents) {
-					// A CompileUnit may not have methods, so guard against this.
-					List<SequencePoint []> list;
-					if (!methodMapping.TryGetValue (documentHandle, out list))
-						list = new List<SequencePoint []> ();
-
-					SourceFileDebugInfo info = new SourceFileDebugInfo (list);
-
-					var document = pdb.GetDocument (documentHandle);
-					info.Hash = pdb.GetBlobBytes (document.Hash);
-					info.HashAlgorithem = pdb.GetGuid (document.HashAlgorithm);
-					info.FileID = 0;
-					info.FullFilePath = pdb.GetString (document.Name);
-
-					fileToSourceFileInfos [info.FullFilePath] = new List<SourceFileDebugInfo> ();
-
-					if (!fileToSourceFileInfos.ContainsKey (Path.GetFileName (info.FullFilePath)))
-						fileToSourceFileInfos [Path.GetFileName (info.FullFilePath)] = new List<SourceFileDebugInfo> ();
-
-					fileToSourceFileInfos [info.FullFilePath].Add (info);
-					fileToSourceFileInfos [Path.GetFileName (info.FullFilePath)].Add (info);
-				}
-			}
-
-			return true;
-		}
-
+			var fileToSourceFileInfos = new Dictionary<string, List<SourceFileDebugInfo>>(StringComparer.InvariantCultureIgnoreCase);
+			sourceFilesDebugInfo[pdbFileName] = fileToSourceFileInfos;
+
+
+			using (var module = Cecil.ModuleDefinition.ReadModule(Path.ChangeExtension(pdbFileName, ".dll")))
+			{
+				var symbolReaderProvider = new Cecil.Cil.DefaultSymbolReaderProvider(false);
+				module.ReadSymbols(symbolReaderProvider.GetSymbolReader(module, pdbFileName));
+
+				var methodMapping = new Dictionary<Document, List<SequencePoint[]>>();
+				foreach (var type in module.Types)
+				{
+					LoadPdbType(type, fileToSourceFileInfos);
+
+					foreach (var nestedType in type.NestedTypes)
+						LoadPdbType(nestedType, fileToSourceFileInfos);
+				}
+			}
+
+			return true;
+		}
+
+		private void LoadPdbType(Cecil.TypeDefinition type, Dictionary<string, List<SourceFileDebugInfo>> fileToSourceFileInfos)
+		{
+			foreach (var method in type.Methods)
+			{
+				var documents = method.DebugInformation.SequencePoints.GroupBy(t => t.Document, t => t, (document, sequencePoints) => new { Document = document, SequencePoints = sequencePoints });

```

Comment:
I'm grouping the documents to be able to iterate document instead of sequence points. It makes the code way more clear to read in the next few lines.